File: codegen/results/v002/qwen3.8_27b_think/20260827_164813/typescript/007_beatles_interview/solution.ts
import * as fs from "fs";
import * as path from "path";

interface Relative {
  FirstName: string;
  LastName: string;
  Relationship: string;
}

interface Person {
  FirstName: string;
  LastName: string;
  Birthday: string;
  Age: number;
  Relatives: Relative[];
}

function parseDate(dateStr: string): { year: number; month: number; day: number } {
  const parts = dateStr.split("/").map(Number);
  return { month: parts[0], day: parts[1], year: parts[2] };
}

function formatDate(dateStr: string): string {
  const { year, month, day } = parseDate(dateStr);
  return `${year}-${String(month).padStart(2, "0")}-${String(day).padStart(2, "0")}`;
}

function calculateAge(birthday: { year: number; month: number; day: number }, reference: { year: number; month: number; day: number }): number {
  let age = reference.year - birthday.year;
  if (reference.month < birthday.month || (reference.month === birthday.month && reference.day < birthday.day)) {
    age--;
  }
  return age;
}

function main(): void {
  const csvPath = path.resolve("input/input.csv");
  const content = fs.readFileSync(csvPath, "utf-8").trim();
  const lines = content.split("\n");
  const headers = lines[0].split(",");
  const rows = lines.slice(1).map((line) => {
    const cols = line.split(",");
    const obj: Record<string, string> = {};
    headers.forEach((h, i) => {
      obj[h.trim()] = (cols[i] || "").trim();
    });
    return obj;
  });

  const referenceDate = { year: 2025, month: 7, day: 1 };

  const result: Person[] = rows.map((row) => {
    const fullName = row["Name"];
    const nameParts = fullName.split(" ");
    const firstName = nameParts[0];
    const lastName = nameParts[nameParts.length - 1];

    const birthday = parseDate(row["Birthday"]);
    const died = row["Died"] !== "null" ? row["Died"] : null;

    let age: number;
    if (died) {
      const deathDate = parseDate(died);
      age = calculateAge(birthday, deathDate);
    } else {
      age = calculateAge(birthday, referenceDate);
    }

    const relatives: Relative[] = [];
    const relMap: Array<{ key: string; relationship: string }> = [
      { key: "Father", relationship: "Father" },
      { key: "Mother", relationship: "Mother" },
      { key: "Brother", relationship: "Brother" },
      { key: "Sister", relationship: "Sister" },
    ];

    for (const { key, relationship } of relMap) {
      const val = row[key];
      if (val && val !== "null") {
        const parts = val.split(" ");
        relatives.push({
          FirstName: parts[0],
          LastName: parts[parts.length - 1],
          Relationship: relationship,
        });
      }
    }

    return {
      FirstName: firstName,
      LastName: lastName,
      Birthday: formatDate(row["Birthday"]),
      Age: age,
      Relatives: relatives,
    };
  });

  process.stdout.write(JSON.stringify(result, null, 2) + "\n");
}

main();

File: codegen/results/v002/qwen3.8_27b_think/20260827_164813/go/007_beatles_interview/solution.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

type Relative struct {
	FirstName    string `json:"FirstName"`
	LastName     string `json:"LastName"`
	Relationship string `json:"Relationship"`
}

type Person struct {
	FirstName string     `json:"FirstName"`
	LastName  string     `json:"LastName"`
	Birthday  string     `json:"Birthday"`
	Age       int        `json:"Age"`
	Relatives []Relative `json:"Relatives"`
}

func parseName(fullName string) (string, string) {
	parts := strings.Split(strings.TrimSpace(fullName), " ")
	firstName := parts[0]
	lastName := parts[len(parts)-1]
	return firstName, lastName
}

func formatDate(dateStr string) string {
	t, _ := time.Parse("1/2/2006", dateStr)
	return t.Format("2006-01-02")
}

func calculateAge(birthStr string, diedStr string) int {
	birth, _ := time.Parse("1/2/2006", birthStr)

	var reference time.Time
	if diedStr != "null" && diedStr != "" {
		reference, _ = time.Parse("1/2/2006", diedStr)
	} else {
		reference = time.Date(2025, 7, 1, 0, 0, 0, 0, time.UTC)
	}

	age := reference.Year() - birth.Year()
	if reference.YearDay() < birth.YearDay() {
		age--
	}
	return age
}

func main() {
	file, err := os.Open("input/input.csv")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading CSV: %v\n", err)
		os.Exit(1)
	}

	var people []Person
	// Skip header row
	for _, record := range records[1:] {
		name := strings.TrimSpace(record[0])
		birthday := strings.TrimSpace(record[1])
		died := strings.TrimSpace(record[2])
		father := strings.TrimSpace(record[3])
		mother := strings.TrimSpace(record[4])
		brother := strings.TrimSpace(record[5])
		sister := strings.TrimSpace(record[6])

		firstName, lastName := parseName(name)

		person := Person{
			FirstName: firstName,
			LastName:  lastName,
			Birthday:  formatDate(birthday),
			Age:       calculateAge(birthday, died),
			Relatives: []Relative{},
		}

		if father != "null" && father != "" {
			fn, ln := parseName(father)
			person.Relatives = append(person.Relatives, Relative{FirstName: fn, LastName: ln, Relationship: "Father"})
		}
		if mother != "null" && mother != "" {
			fn, ln := parseName(mother)
			person.Relatives = append(person.Relatives, Relative{FirstName: fn, LastName: ln, Relationship: "Mother"})
		}
		if brother != "null" && brother != "" {
			fn, ln := parseName(brother)
			person.Relatives = append(person.Relatives, Relative{FirstName: fn, LastName: ln, Relationship: "Brother"})
		}
		if sister != "null" && sister != "" {
			fn, ln := parseName(sister)
			person.Relatives = append(person.Relatives, Relative{FirstName: fn, LastName: ln, Relationship: "Sister"})
		}

		people = append(people, person)
	}

	output, err := json.MarshalIndent(people, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshaling JSON: %v\n", err)
		os.Exit(1)
	}
	fmt.Println(string(output))
}